File: app/src/main/java/com/lence/penguinsvskillerwhales/view/Field.java
//package com.lence.penguinsvskillerwhales.view;
//
//import android.graphics.Point;
//import android.util.Log;
//
//import com.lence.penguinsvskillerwhales.model.Organism;
//import com.lence.penguinsvskillerwhales.model.Penguin;
//import com.lence.penguinsvskillerwhales.utils.Constants;
//
//import java.util.ArrayList;
//import java.util.Collections;
//import java.util.List;
//
//public class Field {
//
//    private int height;
//    private int width;
//    private List<Organism> mLists;
//
//    public Field(int height, int width) {
//        this.height = height;
//        this.width = width;
//        mLists = Collections.synchronizedList(new ArrayList<Organism>(height * width));
//    }
//
//    public void set(int x, int y, Organism organism) {
//        mLists.set(y * width + x, organism);
//    }
//
//    public void set(Point thisPoint, Organism organism) {
//        mLists.set(thisPoint.y * width + thisPoint.x, organism);
//    }
//
//    public Organism get(int x, int y) {
//        Log.e("get", x+" "+y);
//        return mLists.get(y * width + x);
//    }
//
//    public Organism get(Point thisPoint) {
//        return mLists.get(thisPoint.y * width + thisPoint.x);
//    }
//
//    public boolean isInField(int x, int y) {
//        return ((x >= 0) && (x < width) && (y >= 0) && (y < height));
//    }
//
//    public void add(Organism organism) {
//        mLists.add(organism);
//    }
//
//    public static Point getPointByDirection(int x,int y, int direction) {
//        switch (direction) {
//            case Constants.left: {
//                return new Point(x-1,y);
//            }
//            case Constants.right: {
//                return new Point(x+1,y);
//            }
//            case Constants.up: {
//                return new Point(x,y-1);
//            }
//            case Constants.down: {
//                return new Point(x,y+1);
//            }
//            case Constants.left_up: {
//                return new Point(x-1,y-1);
//            }
//            case Constants.right_up: {
//                return new Point(x+1,y-1);
//            }
//            case Constants.left_down: {
//                return new Point(x-1,y+1);
//            }
//            case Constants.right_down: {
//                return new Point(x+1,y+1);
//            }
//        }
//        return new Point(x,y);
//    }
//
//    public int size() {
//        return width * height;
//    }
//
//
//    /*public List<Organism> getmLists() {
//        return mLists;
//    }
//
//    public void setmLists(List<Organism> mLists) {
//        this.mLists = mLists;
//    }*/
//}

File: app/src/main/java/com/lence/penguinsvskillerwhales/view/Field.kt
package com.lence.penguinsvskillerwhales.view

import android.graphics.Point
import com.lence.penguinsvskillerwhales.model.Organism
import com.lence.penguinsvskillerwhales.utils.Constants
import java.util.*

class Field(private val height: Int, private val width: Int ) {
    private var mLists: MutableList<Organism?> = Collections.synchronizedList(ArrayList(height * width))
    operator fun set(x: Int, y: Int, organism: Organism?) {
        mLists[y * width + x] = organism
    }

    operator fun set(thisPoint: Point, organism: Organism?) {
        mLists[thisPoint.y * width + thisPoint.x] = organism
    }

    operator fun get(x: Int, y: Int): Organism? {
        //Log.e("get", x.toString() + " " + y)
        return mLists[y * width + x]
    }

    operator fun get(thisPoint: Point): Organism? {
        return mLists[thisPoint.y * width + thisPoint.x]
    }

    fun isInField(x: Int, y: Int): Boolean {
        return x in 0..(width - 1) && y in 0..(height-1)
    }

    fun add(organism: Organism?) {
        mLists.add(organism)
    }
    companion object {
        @JvmStatic
        fun getPointByDirection(x: Int, y: Int, direction: Int): Point {
            when (direction) {
                Constants.left -> {
                    return Point(x - 1, y)
                }
                Constants.right -> {
                    return Point(x + 1, y)
                }
                Constants.up -> {
                    return Point(x, y - 1)
                }
                Constants.down -> {
                    return Point(x, y + 1)
                }
                Constants.left_up -> {
                    return Point(x - 1, y - 1)
                }
                Constants.right_up -> {
                    return Point(x + 1, y - 1)
                }
                Constants.left_down -> {
                    return Point(x - 1, y + 1)
                }
                Constants.right_down -> {
                    return Point(x + 1, y + 1)
                }
            }
            return Point(x, y)
        }
    }

    fun size(): Int {
        return width * height
    }

}